packages/calling-stateful-client/src/DeviceManager.test.ts

```diff
@@ -0,0 +1,233 @@
+
+    expect(client.getState().deviceManager.selectedMicrophone).not.toBeDefined();
+    manager.selectedMicrophone = microphoneWithName('firstMicrophone');
+    manager.emit('selectedMicrophoneChanged', {});
+    expect(
+      await waitWithBreakCondition(() => client.getState().deviceManager.selectedMicrophone?.name === 'firstMicrophone')
+    ).toBe(true);
+  });
+
+  test('should proxy getSpeakers and update speakers in state', async () => {
+    const client = createStatefulCallClientWithDeviceManager(
+      createMockDeviceManagerWithSpeakers([speakerWithName('speaker')])
+    );
+
+    const speakers = await (await client.getDeviceManager()).getSpeakers();
+    expect(speakers.length).toBe(1);
+    expect(speakers[0].name).toBe('speaker');
+
+    expect(client.getState().deviceManager.speakers.length).toBe(1);
+    expect(client.getState().deviceManager.speakers[0].name).toBe('speaker');
+  });
+
+  test('should proxy selectSpeaker and update selectedSpeaker in state', async () => {
+    const speaker = speakerWithName('speaker');
+    const client = createStatefulCallClientWithDeviceManager(createMockDeviceManagerWithSpeakers([speaker]));
+
+    expect(client.getState().deviceManager.selectedSpeaker).not.toBeDefined();
+
+    await (await client.getDeviceManager()).selectSpeaker(speaker);
+    expect(client.getState().deviceManager.selectedSpeaker).toBeDefined();
+    expect(client.getState().deviceManager.selectedSpeaker?.name).toBe('speaker');
+  });
+
+  test('should detect selectedSpeakerChanged update selectedSpeaker in state using deviceManager', async () => {
+    const manager = createMockDeviceManagerWithSpeakers([speakerWithName('speaker')]);
+    const client = createStatefulCallClientWithDeviceManager(manager);
+    // FIXME? Device manager is only created on request, so state is not updated unless device manager is created.
+    await client.getDeviceManager();
+
+    expect(client.getState().deviceManager.selectedSpeaker).not.toBeDefined();
+    manager.selectedSpeaker = speakerWithName('speaker');
+    manager.emit('selectedSpeakerChanged', {});
+    expect(
+      await waitWithBreakCondition(() => client.getState().deviceManager.selectedSpeaker?.name === 'speaker')
+    ).toBe(true);
+  });
+
+  test('should detect audioDevicesUpdated update microphones/speakers in state using deviceManager', async () => {
+    const manager = createMockDeviceManagerWithSpeakers(
+      [speakerWithName('secondSpeaker')],
+      createMockDeviceManagerWithMicrophones([microphoneWithName('firstMicrophone')])
+    );
+    const client = createStatefulCallClientWithDeviceManager(manager);
+    // FIXME? Device manager is only created on request, so state is not updated unless device manager is created.
+    await client.getDeviceManager();
+
+    manager.emit('audioDevicesUpdated', {});
+    expect(
+      await waitWithBreakCondition(() => client.getState().deviceManager.microphones[0]?.name === 'firstMicrophone')
```

Comment:
I forget - what's the reason we need to wait literal milliseconds here? Is there a way to mock timers to avoid it?
(This isn't a change of this PR I know so this is not a blocking comment - just want to get rid of arbitrary waits 😅 )

Reply:
Renaming suggestions welcome. This isn't actually a wait. This is more a wait-till-condition-becomes-true-or-timeout.
This is needed so that we can assert conditions that become true once asynchronous operations complete.
So, when the test is passing, the expected wait is ~0 second.

Reply:
Incidentally, this PR fixes a bug where this helper was always waiting 100 milliseconds before checking for the first time.

Reply:
I guess amortized 0 seconds is acceptable :P (though my worry is it sets a precedence where the codebase feels its acceptable to have wait-x-milliseconds conditions)

Reply:
Realized after commenting here that the first check is likely going to always fail -- the async operations haven't had a chance to run.
So... we'll still waste 100 milliseconds.

But hey, it's just the test machine's CPU. After some experimentation, 10ms is a better wait time on my workstation. The test runtime goes down from ~20 seconds to ~2 seconds.

So... there.